### DISCUSSION TITLE
When is cartesian product supported?

### DISCUSSION THREAD

**Discussion by pjljvandelaar** [junit-team/junit-framework#3088] [Q&A] [answered]
Dear Junit community,

I am missing the feature `Cartesian Product` that has been researched by Junit Pioneer.
Is integration in JUnit planned?
And if so, when will ite be integrated?

Thanks in advance for the answers!

Pierre

**Comment by portyu9** [comment, marked as answer]
Hi @pjljvandelaar 

As of JUnit 6.1.3, Cartesian-product parameterization has **not been integrated into JUnit Jupiter**, and there is no announced integration date.

The current JUnit Pioneer API is `@CartesianTest`, for example:

```java
@CartesianTest
void test(
        @CartesianTest.Values(ints = {1, 2}) int x,
        @CartesianTest.Values(strings = {"A", "B"}) String value) {
    // executed for all 4 combinations
}
```

So JUnit Pioneer remains the straightforward solution today.

One version-related detail: older Pioneer versions used `@CartesianProductTest`. That API was deprecated and subsequently removed; the current API is `@CartesianTest`.

If you prefer not to depend on Pioneer, standard Jupiter can also do it with:

```java
@ParameterizedTest
@MethodSource("arguments")
void test(int x, String value) {
    ...
}
```

where the `arguments()` factory generates the Cartesian product itself.

So there is still no native Jupiter annotation equivalent to Pioneer's `@CartesianTest`; use Pioneer or generate the combinations through an argument source.

Please mark this answer as accepted if it helped, thank you.
